device: push settings when device requests software reconfiguration

pfps opened the pull request:
Fixes #1380

viniciusbm commented:
This is interesting. I think this update can also fix an **unrelated** issue that may affect devices that are not connected via Bluetooth. 

About a year ago, my K400+ keyboard was randomly forgetting all its settings, and sometimes it happened less than a second after the last keystroke, so it wasn't after going idle or being powered off.

After some investigation, I noticed that the device requested reconfiguration whenever it forgot the settings, and [my workaround](https://github.com/viniciusbm/Solaar/commit/8c7c36e8777040cd350e00629e4b986f78b65984) was very similar to what this pull request does. I didn't create an issue or PR at the time because I assumed it was a hardware fault, and a few months later I replaced the batteries and the problem was apparently solved (but the old ones were still at 50%, so I'm not sure this was the reason).

Since the documentation for 0x1D4B does not say that software reconfiguration is only needed under specific circumstances, I think fulfilling the reconfiguration requests unconditionally as done by this PR is a good idea.

pfps commented:
@viniciusbm It would be useful to obtain the notification message that your keyboard sent out.  This would show more about why they keyboard is requesting reconfiguration.  It could be because of some hardware malfunction, or battery glitch, or communications glitch.

pfps commented:
@viniciusbm Could you test out this PR?

viniciusbm commented:
> @viniciusbm It would be useful to obtain the notification message that your keyboard sent out. This would show more about why they keyboard is requesting reconfiguration. It could be because of some hardware malfunction, or battery glitch, or communications glitch. 

I don't have the full logs anymore, but I found the relevant log lines in [the draft](https://drive.google.com/file/d/1Am2S3OIG7NgLyes9ZwXPuALq1BuQCg23/view?usp=sharing) I wrote in December 2020 with the intention of opening an issue:

<details>

From my draft, verbatim:

> _I've noticed that my K400+ keyboard sporadically "forgets" its settings **while it's being used**. I ran Solaar with `-dd` and redirected stderr to a file for a several days. Then, I examined the logs immediately after I noticed that the settings had been lost, and as I expected, there are no entries that might indicate that the device went idle or powered off._

> _Before and after the following lines, there were lines representing key presses just a few milliseconds earlier/later, which confirms that this happens while I was using the keyboard. All of the occurrences follow exactly this pattern._

```
08:53:14,500    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (15) => r[11 01 0400 01010100000000000000000000000000]
08:53:14,500    DEBUG [ReceiverListener:hidraw0] logitech_receiver.notifications: wireless status: Notification(11,1,04,00,01010100000000000000000000000000)
08:53:14,500     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)>: paired online, {'BATTERY LEVEL': 90, 'LINK ENCRYPTED': True, 'BATTERY STATUS': NamedInt(0, 'discharging'), 'BATTERY NEXT LEVEL': 50, 'BATTERY VOLTAGE': None, 'BATTERY CHARGING': False, 'ERROR': None} (1) powered on
08:53:14,501    DEBUG [MainThread] solaar.ui: status changed: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (NONE) powered on
08:53:14,501    DEBUG [MainThread] solaar.ui.icons: battery icon for 90:False = battery-full
08:53:14,502    DEBUG [MainThread] solaar.ui.tray: picked device with lowest battery: ('/dev/hidraw0', 1, 'Wireless Touch Keyboard K400 Plus', {'BATTERY LEVEL': 90, 'LINK ENCRYPTED': True, 'BATTERY STATUS': NamedInt(0, 'discharging'), 'BATTERY NEXT LEVEL': 50, 'BATTERY VOLTAGE': None, 'BATTERY CHARGING': False, 'ERROR': None})
08:53:14,502    DEBUG [MainThread] solaar.ui.icons: battery icon for 90:False = battery-full
08:53:14,503    DEBUG [MainThread] solaar.ui.icons: battery icon for 90:False = battery-full
08:53:14,503    DEBUG [MainThread] solaar.ui.icons: battery icon for 90:False = battery-full
```

> _At first, I thought this might be caused by some hardware fault; however, the first line that's not generated by key presses **always** refers to `WirelessDeviceStatus` with bytes `01 01 01`, and  [the documentation](https://docs.google.com/document/d/1dItDNvUbByN3RXPQnJotv2WwhGxovXogLVHsBm-womQ/edit?usp=sharing) does **not** say that software reconfiguration is only needed when the device was powered off._

> _Hence, I think Solaar should push settings whenever the device asks for reconfiguration, even if it was already powered on._

</details>

The problem hasn't occurred since I replaced the batteries in February, but I'll record the logs if it ever comes up again.

viniciusbm commented:
> @viniciusbm Could you test out this PR?

I cannot confirm that the changes work because I don't have any devices currently affected by the issue it solves, but maybe there's something I can test. Are there any specific tests I can do?

pfps commented:
@viniciusbm The changes affect lots of devices.  Any device that has     WIRELESS_DEVICE_STATUS=0x1D4B, which includes most modern Logitech devices, is potentially affected.

Run this PR as `bin/solaar -dd`.
For each of your devices, turn them off and on again.  Do this both quickly and slowly.
The ideal situation is that the device disappears from the Solaar main window if it is connected via Bluetooth or a USB cable or becomes inactive if it is connected to a receiver.  Then when the device is turned on Solaar configures it, but only once.   The debug output of Solaar prints lines containing `pushing device settings` at the beginning of configuring a device so there should be exactly one of these lines each time a device is turned on.  If there are two of these lines its a problem but not a severe one.  If there is no line, then it is a severe problem, so please post the output from Solaar from before the device was turned off to after the device was turned on.

pfps commented:
@viniciusbm With this PR Solaar should reconfigure the device whenever it receives a notification like: `wireless status: Notification(11,1,04,00,01010100000000000000000000000000)` so it looks as if this PR fixes the problem you encountered.

viniciusbm commented:
> For each of your devices, turn them off and on again. Do this both quickly and slowly.
> The ideal situation is that the device [...] becomes inactive if it is connected to a receiver. Then when the device is turned on Solaar configures it, but only once. The debug output of Solaar prints lines containing `pushing device settings` at the beginning of configuring a device so there should be exactly one of these lines each time a device is turned on.

I've just tested using my only Logitech device connected to a Unifying Receiver, and the program behaves exactly like you described. 

Slowly (off at :30, on at :40, off at :50, on at :00):
[slow-off-on.txt](https://github.com/pwr-Solaar/Solaar/files/7740187/slow-off-on.txt)

Quickly (off-on at :30, :40 and :50):
[fast-off-on.txt](https://github.com/pwr-Solaar/Solaar/files/7740188/fast-off-on.txt)

Lines with `pushing device settings`, as expected: 

```
$ grep 'pushing device' {slow,fast}-off-on.txt
slow-off-on.txt:21:27:24,050     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: None) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=None)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys=None)>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures=None)>, <Setting([feature:multiple range] K400 Plus:gesture2-params=None)>, <Setting([feature:map choice] K400 Plus:divert-keys=None)>]
slow-off-on.txt:21:27:40,411     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: True) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=True)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys={'1': False, '16': False, '4': False, '8': False})>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures={'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': False})>, <Setting([feature:multiple range] K400 Plus:gesture2-params={'4': {'scale': 256}})>, <Setting([feature:map choice] K400 Plus:divert-keys={'1': 0, '110': 1, '145': 1, '149': 1, '183': 1, '184': 1, '2': 0, '3': 0, '34': 1, '4': 1, '40': 1, '5': 1, '6': 1, '62': 1, '64': 0, '79': 1, '8': 1, '84': 1})>]
slow-off-on.txt:21:28:00,364     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: True) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=True)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys={'1': False, '16': False, '4': False, '8': False})>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures={'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': False})>, <Setting([feature:multiple range] K400 Plus:gesture2-params={'4': {'scale': 256}})>, <Setting([feature:map choice] K400 Plus:divert-keys={'1': 0, '110': 1, '145': 1, '149': 1, '183': 1, '184': 1, '2': 0, '3': 0, '34': 1, '4': 1, '40': 1, '5': 1, '6': 1, '62': 1, '64': 0, '79': 1, '8': 1, '84': 1})>]
fast-off-on.txt:21:28:24,117     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: None) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=None)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys=None)>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures=None)>, <Setting([feature:multiple range] K400 Plus:gesture2-params=None)>, <Setting([feature:map choice] K400 Plus:divert-keys=None)>]
fast-off-on.txt:21:28:30,714     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: True) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=True)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys={'1': False, '16': False, '4': False, '8': False})>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures={'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': False})>, <Setting([feature:multiple range] K400 Plus:gesture2-params={'4': {'scale': 256}})>, <Setting([feature:map choice] K400 Plus:divert-keys={'1': 0, '110': 1, '145': 1, '149': 1, '183': 1, '184': 1, '2': 0, '3': 0, '34': 1, '4': 1, '40': 1, '5': 1, '6': 1, '62': 1, '64': 0, '79': 1, '8': 1, '84': 1})>]
fast-off-on.txt:21:28:40,707     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: True) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=True)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys={'1': False, '16': False, '4': False, '8': False})>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures={'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': False})>, <Setting([feature:multiple range] K400 Plus:gesture2-params={'4': {'scale': 256}})>, <Setting([feature:map choice] K400 Plus:divert-keys={'1': 0, '110': 1, '145': 1, '149': 1, '183': 1, '184': 1, '2': 0, '3': 0, '34': 1, '4': 1, '40': 1, '5': 1, '6': 1, '62': 1, '64': 0, '79': 1, '8': 1, '84': 1})>]
fast-off-on.txt:21:28:50,823     INFO [ReceiverListener:hidraw1] logitech_receiver.status: <Device(1,404D,Wireless Touch Keyboard K400 Plus,6DC5C81C)> (was_active: True) pushing device settings [<Setting([feature:toggle] K400 Plus:fn-swap=True)>, <Setting([feature:multiple toggle] K400 Plus:disable-keyboard-keys={'1': False, '16': False, '4': False, '8': False})>, <Setting([feature:multiple toggle] K400 Plus:gesture2-gestures={'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': False})>, <Setting([feature:multiple range] K400 Plus:gesture2-params={'4': {'scale': 256}})>, <Setting([feature:map choice] K400 Plus:divert-keys={'1': 0, '110': 1, '145': 1, '149': 1, '183': 1, '184': 1, '2': 0, '3': 0, '34': 1, '4': 1, '40': 1, '5': 1, '6': 1, '62': 1, '64': 0, '79': 1, '8': 1, '84': 1})>]
```

PS: In my tests, I added `was_active` to the log: 
```python
_log.info('%s (was_active: %r) pushing device settings %s', d, was_active, d.settings)
```

PixlRainbow commented:
I have just tested this with my MX Master 2S, and it does seem to properly re-apply the rules on reconnect now.

pfps commented:
@PixlRainbow Can you run Solaar as `bin/solaar -dd` with the remaining problematic case and attach the output and the Solaar main window with the detailed info box showing?

PixlRainbow commented:
![image](https://user-images.githubusercontent.com/44422178/146675689-3ee165df-c17c-46de-a7f7-673112267091.png)
(^ it looks the same both before and after the mouse is connected)
[solaardd_deferconnect.log](https://github.com/pwr-Solaar/Solaar/files/7740827/solaardd_deferconnect.log)

pfps commented:
Thanks.  What appears to be happening is that Solaar sees the device, thinks it is active, and tries to write to it but the device does not respond.  So Solaar thinks it is a old device and sets a flag accordingly, which then prevents some later code from working.  I've fixed this and updated the PR.  

Please try again with the new version of the PR.

PixlRainbow commented:
When 2b6b8f9 is started with the mouse disconnected, there is a delay of 30 seconds before the UI window appears, and a further delay of 5 seconds before the interface is populated.  
After the mouse is connected, the battery information is updated correctly, but there are no fields in the settings area and user-defined rules don't seem to be applied.

pfps commented:
Please provide output of `bin/solaar -dd`

PixlRainbow commented:
[solaardd_deferconnect2.log](https://github.com/pwr-Solaar/Solaar/files/7740982/solaardd_deferconnect2.log)

pfps commented:
Thanks.  I see that the most-recent change had some bad effects.

PixlRainbow commented:
I also notice that solaar no longer provides notifications of reconnects, regardless of whether solaar was started with the mouse connected or disconnected. Did you do that too?

pfps commented:
The powered-on notifications have been removed.  They didn't work correctly in many cases and people were complaining that they were intrusive.

pfps commented:
It appears that there are two modes that BlueTooth devices can work in.  In the mode I've seen, the BlueTooth connection drops as soon as the device is turned off and the device is active immediately after it is turned on.  In your mode the BlueTooth connection persists and the device is not active immediately after it is turned on.  In my mode the device does not produce a wireless status notification when it is turned on.  In your mode the device does produce this notification.

I made a couple of changes that should work for both modes.  Please grab the current version of this PR and run your both your use cases again.  If one doesn't work correctly run it again with `bin/solaar -dd` and post the output.  If they both work just respond that things are working.

PixlRainbow commented:
Starting Solaar with the mouse disconnected and then connecting it later still does not work. Battery status is **not updated** and detail info is **not populated** in the UI. Additionally there is a delay of 5-10 seconds before the GUI window appears.
![image](https://user-images.githubusercontent.com/44422178/146682026-b7731be1-9fc5-4304-bd67-677e2430725d.png)
[solaardd_deferconnect3.log](https://github.com/pwr-Solaar/Solaar/files/7741120/solaardd_deferconnect3.log)

pfps commented:
The startup delay is pretty much unavoidable.  Solaar is trying to communicate with the device and is not receiving a response to it has to wait for its timeout to expire.

But there is still something wrong going on.  I'm going to put in some extra debug stuff for you to run.

pfps commented:
Please download the new version of this PR, as described in #1380 and try the deferconnect situation again.

pfps commented:
Please download this PR again, as described in #1380 and try the situation with the mouse initially disconnected.

PixlRainbow commented:
As of b780395 mouse is now detected and all settings and rules are applied upon connection

Something interesting I noticed when turning on my PC today: the release version 1.1.0 will detect the mouse if it is connected after startup -- but only if it is the first time said mouse is connected to the system since OS boot.  
It appears that before that first connection to the OS, the mouse is considered to be completely inactivated, and after that initial connection, the Bluetooth "session" persists across disconnects and reconnects.

pfps commented:
What appears to be happening is that on your system Bluetooth connections are not torn down as soon as the mouse is turned off.   Solaar is listening to the connection but doesn't see anything happening so it doesn't see anything to indicate that the mouse has been turned off.  (It would be nice to find out how your system's drivers notice that the mouse has been turned off.)  

When Solaar is started with the mouse inactive but the Bluetooth connection active the first thing that Solaar does is try to send a command to the mouse.  As the mouse is inactive it does not respond to the command and Solaar waits until its timeout is reached.  This delay is unavoidable (unless there is some other way for Solaar to determine that the mouse is inactive).   What was happening was that Solaar then thought that the mouse was an old-style  mouse so when the mouse became active and sent a new-style notification Solaar wasn't prepared to handle it.  The earlier changes in this PR meant that Solaar no longer thought that the mouse was an old-style mouse but had the side effect that Solaar thought the mouse was permanently inactive.  The last change means that Solaar also realizes that the mouse is active.

So things appear to be good now except for the startup delay.  Fixing this would require a different way of communicating with the mouse and would have to be done by someone with the kind of system you are running.
